atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraAbsentHostTracker.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra;
+
+import com.google.common.collect.ImmutableSet;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.net.InetSocketAddress;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.immutables.value.Value;
+
+public final class CassandraAbsentHostTracker {
+    private static final SafeLogger log = SafeLoggerFactory.get(CassandraAbsentHostTracker.class);
+
+    private final int absenceLimit;
+    private final Map<InetSocketAddress, PoolAndCount> absentHosts;
+
+    public CassandraAbsentHostTracker(int absenceLimit) {
+        this.absenceLimit = absenceLimit;
+        this.absentHosts = new HashMap<>();
+    }
+
+    public synchronized Optional<CassandraClientPoolingContainer> returnPool(InetSocketAddress host) {
+        return Optional.ofNullable(absentHosts.remove(host)).map(PoolAndCount::container);
+    }
+
+    public synchronized void trackAbsentHost(InetSocketAddress host, CassandraClientPoolingContainer pool) {
+        absentHosts.putIfAbsent(host, PoolAndCount.of(pool));
+    }
+
+    public synchronized Set<InetSocketAddress> incrementAbsenceAndRemove() {
+        return cleanupAbsentHosts(ImmutableSet.copyOf(absentHosts.keySet()));
```

Comment:
this probably _is_ needed to avoid ConcurrentModificationException stuff, but is a bit of a dance